geoips/plugins/modules/readers/generic_satpy.py

```diff
@@ -0,0 +1,145 @@
+dictionary's keys should be strings representing the name of each group and the
+values should be a list of channel names that have the same resolution. The name
+of each group will be used as the keys of the output dictionary and the list of
+channels will be grouped together as a Dataset stored as the associated value to
+the key.  So for example the following channel_groups JSON Object could be used
+to read a selection of visible and IR GOES16 data:
+"channel_groups":{"0.5km":["C02"], "2km":["C13", "C14"]}
+
+While Satpy provides a very large number of readers, a very small number of them
+have been used with this plugin.  This plugin should be considered experimental
+and is not guaranteed to work with all of Satpy's readers.
+"""
+
+import logging
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import dask
+import pyresample.geometry
+import satpy
+import xarray as xr
+
+log = logging.getLogger(__name__)
+
+interface = "readers"
+family = "standard"
+name = "generic_satpy"
+
+
+DEFAULT_COORD_CHUNK = 4096
+
+
+def call(
+    fnames: List[str],
+    source_name: str,
+    satpy_reader: str,
+    channel_groups: Dict[str, List[str]],
+    platform_name: Optional[str] = None,
+    metadata_only: bool = False,
+    chans: Optional[List[str]] = None,
+    area_def: Optional[pyresample.geometry.AreaDefinition] = None,
+    self_register: bool = False,
+    roi: float = 3000,
+    satpy_reader_kwargs: Optional[Dict[str, Any]] = None,
+    calibration: Union[List[str], str] = "*",
+    resolution: Union[List[float], float] = "*",
+    polarization: Union[List[str], str] = "*",
+    level: Union[List[float], str] = "*",
+    modifiers: Union[List[str], str] = "*",
+    load_kwargs: Optional[Dict[str, Any]] = None,
+    force_compute: bool = True,
+    coord_chunks: int = DEFAULT_COORD_CHUNK,
+) -> Dict[str, xr.DataArray]:
+    if platform_name is None:
+        platform_name = satpy_reader
+
+    scene = satpy.Scene(
+        filenames=fnames,
+        reader=satpy_reader,
+        reader_kwargs=satpy_reader_kwargs,
+    )
+
+    all_channels = []
+    for channel_names in channel_groups.values():
+        all_channels.extend(channel_names)
+
+    if load_kwargs is None:
+        load_kwargs = {}
+
+    scene.load(
+        all_channels,
+        calibration=calibration,
+        resolution=resolution,
+        polarization=polarization,
+        level=level,
+        modifiers=modifiers,
+        **load_kwargs,
+    )
+
+    output_datasets = {}
+    for group_name, channel_names in channel_groups.items():
+        lons, lats = (None, None)
+        is_first_channel = True
+        data_dict = {}
+        for channel_name in channel_names:
+            channel = scene[channel_name]
+            if is_first_channel:
+                is_first_channel = False
+                lons, lats = _compute_coords(channel, coord_chunks)
+
+            if force_compute and not metadata_only:
+                channel = channel.compute()
```

Comment:
Your comment about making sure that lons and lats are computed at the same time made me wonder if this channel compute should be somehow computed later so that all datasets/products are computed at the same time. In most cases for datasets coming straight from a reader's file this won't matter. However, in cases where there are shared dask computations then doing the compute all together could be a big deal. I *think* you can pass any python native data structure to `dask.array.compute` and it will recurse through it to find dask arrays to compute. So you could take `output_datasets` and do `return dask.array.compute(output_datasets)`. Again, I think, it should be tested.

Reply:
I'll give this a try.  If this doesn't work, then it may be possible to write some simple alternative logic that would accomplish the same thing.

Reply:
At the moment, I'm going to leave this one as a future optimization/improvement.